### PR TITLE
ORV2-1504: Applying for no-fee permits

### PR DESCRIPTION
**Pull request by zgong-gov**
# Description

- Allow no-fee permit types to be used for all permits
- Application review page and shopping cart items show $0 for fee summary when no-fee flag enabled, and normal fees when flag disabled
- Amend refund works normally, except that permits with net amount of $0 gets refunded $0
- Void refund works normally, except that permits with net amount of $0 gets refunded $0

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Set the no-fee flag in special authorizations page, and create a new application. Verify that the fee summary on the application review and shopping cart page (and cart items) all show fee summary of $0
- [x] Disable the no-fee flag in special authorizations page, and create a new application. Verify that the fee summary on the application review and shopping cart page (and cart items) all show correctly calculated (non-zero) fees
- [x] Amend a permit several times by reducing its duration or changing other values, and observe that for any permit that was initially purchased with non-zero dollars, regardless of how the no-fee flag is toggled during this amend process, the proper refund amount will always be refunded (ie. reducing 90 day duration to 60 days for TROS always refunds $30, regardless of no-fee flag)
- [x] Amend a permit paid with $0 several times by reducing its duration or changing other values, and observe that the refund amount is always $0, regardless of what values were changed for amend, and how the no-fee flag is toggled during this amend process
- [x] Void a permit paid with $0 and observe that the refund amount is always $0, regardless of what the no-fee flag status is in, nor the duration of the permit
- [x] Void a non-zero dollar permit, and observe that the original normal void behaviour still works as expected, regardless of the no-fee flag status

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

## Further comments

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://onroutebc-1595-frontend.apps.silver.devops.gov.bc.ca)
- [Vehicles](https://onroutebc-1595-vehicles.apps.silver.devops.gov.bc.ca/api)
- [Dops](https://onroutebc-1595-dops.apps.silver.devops.gov.bc.ca/api)
- [Policy](https://onroutebc-1595-policy.apps.silver.devops.gov.bc.ca/api)
- [Scheduler](https://onroutebc-1595-scheduler.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/onroutebc/actions/workflows/analysis.yml)

After merge, new images are promoted to:
- [Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge.yml)

### CODE REVIEW

File: vehicles/src/common/helper/permit-fee.helper.ts [hunk 1]
```diff
@@ -25,7 +25,11 @@ import { ApplicationStatus } from '../enum/application-status.enum';
  * @throws {NotAcceptableException} If the duration is invalid for TROS permit type.
  * @throws {BadRequestException} If the permit type is not recognized.
  */
-export const permitFee = (application: Permit, oldAmount?: number): number => {
+export const permitFee = (
+  application: Permit,
+  isNoFee?: boolean,
+  oldAmount?: number|undefined,
```
Comment (praju-aot) [inline vehicles/src/common/helper/permit-fee.helper.ts]: `Nullable<number>`

File: vehicles/src/common/helper/permit-fee.helper.ts [hunk 2]
```diff
@@ -25,7 +25,11 @@ import { ApplicationStatus } from '../enum/application-status.enum';
  * @throws {NotAcceptableException} If the duration is invalid for TROS permit type.
  * @throws {BadRequestException} If the permit type is not recognized.
  */
-export const permitFee = (application: Permit, oldAmount?: number): number => {
+export const permitFee = (
+  application: Permit,
+  isNoFee?: boolean,
```
Comment (praju-aot) [inline vehicles/src/common/helper/permit-fee.helper.ts]: `Nullable<boolean>`

File: vehicles/src/common/helper/permit-fee.helper.ts [hunk 3]
```diff
@@ -147,14 +153,27 @@ export const currentPermitFee = (
   duration: number,
   pricePerTerm: number,
   allowedPermitTerm: number,
-  oldAmount?: number,
+  oldAmount?: number|undefined,
```
Comment (praju-aot) [inline vehicles/src/common/helper/permit-fee.helper.ts]: `Nullable<number>`

File: vehicles/src/common/helper/permit-fee.helper.ts [hunk 4]
```diff
@@ -147,14 +153,27 @@ export const currentPermitFee = (
...
   permitStatus?: ApplicationStatus,
+  isNoFee?: boolean,
```
Comment (praju-aot) [inline vehicles/src/common/helper/permit-fee.helper.ts]: `Nullable<boolean>`

File: vehicles/src/common/helper/permit-fee.helper.ts [hunk 5]
```diff
@@ -147,14 +153,27 @@ export const currentPermitFee = (
   duration: number,
   pricePerTerm: number,
   allowedPermitTerm: number,
-  oldAmount?: number,
+  oldAmount?: number|undefined,
   permitStatus?: ApplicationStatus,
```
Comment (praju-aot) [inline vehicles/src/common/helper/permit-fee.helper.ts]: `Nullable<ApplicationStatus>`

File: vehicles/src/common/helper/permit-fee.helper.ts
Comment (praju-aot) [inline vehicles/src/common/helper/permit-fee.helper.ts:1]: @gchauhan-aot Please confirm that application is not persisting any amount as -ve in the DB

**Review by praju-aot — CHANGES_REQUESTED** [review]

File: vehicles/src/common/helper/permit-fee.helper.ts
Comment (gchauhan-aot) [reply vehicles/src/common/helper/permit-fee.helper.ts:1]: We are not persisting -ve amounts in DB.

File: vehicles/src/common/helper/permit-fee.helper.ts (hunk 2 above)
Comment (gchauhan-aot) [reply vehicles/src/common/helper/permit-fee.helper.ts]: done

File: vehicles/src/common/helper/permit-fee.helper.ts (hunk 1 above)
Comment (gchauhan-aot) [reply vehicles/src/common/helper/permit-fee.helper.ts]: done

File: vehicles/src/common/helper/permit-fee.helper.ts (hunk 3 above)
Comment (gchauhan-aot) [reply vehicles/src/common/helper/permit-fee.helper.ts]: done

File: vehicles/src/common/helper/permit-fee.helper.ts (hunk 5 above)
Comment (gchauhan-aot) [reply vehicles/src/common/helper/permit-fee.helper.ts]: done

File: vehicles/src/common/helper/permit-fee.helper.ts (hunk 4 above)
Comment (gchauhan-aot) [reply vehicles/src/common/helper/permit-fee.helper.ts]: done

**Review by praju-aot — APPROVED** [review]
+1

**Pull request merged**